Add macro to generate register structs, and apply it to nrf52/clock (Fix #1400)

gendx opened the pull request:
### Pull Request Overview

This pull request adds helper macros to generate register structs, as described in https://github.com/tock/tock/issues/1400. As an example, `chips/nrf52/src/clock.rs` is migrated to use these macros.

The `register_structs` macro does two things:
- Generate the register structs, in particular computing the length of padding fields.
- Create unit tests, checking that each field indeed takes the size between its offset and the next field's offset. These tests require https://github.com/tock/tock/pull/1393 to run (one can run them manually with e.g. `cd chips/nrf52 && CI=true TOCK_KERNEL_VERSION=ci_test cargo test`, although you need to apply the fixes from https://github.com/tock/tock/pull/1393).

Some parts are still not complete, as I'm looking for feedback first.

### Testing Strategy

This pull request was tested by:
- Running `make ci-travis`.
- Running `cargo test` for the nrf52 chip, as described in https://github.com/tock/tock/pull/1393.
- Checking the offsets in the data sheet for the clock at https://infocenter.nordicsemi.com/pdf/nRF52840_OPS_v0.5.pdf (section 17.3).
- Manually double-checking the paddings.
- Flashing the firmware on a nRF52840-DK board and running a simple app (blink).

### TODO or Help Wanted

This pull request still needs:

- [x] Updated `svd2regs.py` script to generate the macro instead of a manual struct.
- [x] Updated documentation for the tock-register-interface library.

I'm waiting for feedback before adding these.

Additionally:

- [x] The generated tests currently require `use super::super::*;` due to putting them in a nested module. Wildcard imports are currently prohibited by Tock style guide. I added an exception for this case.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

ppannuto commented:
This is awesome, +1 here as well once there's documentation added to the register interface.

One maybe-nitpick on the syntax: is it weird to not have commas after each entry? This looks like a list-like thing, my intuition expects commas. While I find it perfectly readable with just newlines as-is, I think commas would be more consistent with rust syntax as a whole?

gendx commented:
The extra commas make sense. I also updated the syntax to mark the end with `@END`, which stands out more and doesn't conflict with an identifier.

Added the documentation for the register interface.

So the remaining points are `svd2regs.py` and the fact that unit tests currently require a wildcard import (although they won't run on CI until https://github.com/tock/tock/pull/1393 is merged).

ia0 commented:
I'm new to embedded programming, but I wonder why use a `struct` for registers in Rust. The API we want is given a register name, get access to that register. This can indeed be encoded with the field offsets of a repr(C) struct, but it can also be encoded using constant functions (producing the same assembly). Those fake structs don't exist, we never allocate them. They are just a way to encode an offset and provide a way to add it to a base address. This can be done with a constant function too.

To illustrate consider this macro to generate registers:
```rust
macro_rules! define_registers {
    (struct $Name: ident { $($name: ident: $type: ty = $offset: expr),*$(,)? }) => {
        #[derive(Clone, Copy)]
        struct $Name { base_address: usize }
        impl $Name {
            $(const fn $name(&self) -> StaticRef<$type> {
                unsafe { StaticRef::new((self.base_address + $offset) as *const $type) }
            })*
        }
        #[test]
        #[allow(non_snake_case)]
        fn $Name() {
            struct Field { name: &'static str, offset: usize, size: usize }
            let mut fields = [$(Field {
                name: stringify!($name),
                offset: $offset,
                size: core::mem::size_of::<$type>()
            }),*];
            fields.sort_by_key(|field| field.offset);
            for window in fields.windows(2) {
                let (prev, next) = (&window[0], &window[1]);
                assert!(prev.offset + prev.size <= next.offset,
                        "{} and {} overlap in {}",
                        prev.name, next.name, stringify!($Name));
            }
        }
    };
}
```

When given the following input:
```rust
define_registers! {
    struct ClockRegisters {
        tasks_hfclkstart: WriteOnly<u32, Control::Register> = 0x000,
        tasks_hfclkstop: WriteOnly<u32, Control::Register> = 0x004,
        tasks_lfclkstart: ReadWrite<u32, Control::Register> = 0x008,
        ...
        lfclksrccopy: ReadOnly<u32, LfClkSrcCopy::Register> = 0x41C,
        lfclksrc: ReadWrite<u32, LfClkSrc::Register> = 0x518,
        ctiv: ReadWrite<u32, Ctiv::Register> = 0x538,
        traceconfig: ReadWrite<u32, TraceConfig::Register> = 0x55C,
    }
}
```

It produces the following output:
```rust
struct ClockRegisters { base_address: usize }
impl ClockRegisters {
    const fn tasks_hfclkstart(&self) -> StaticRef<WriteOnly<u32, Control::Register>> {
        unsafe { StaticRef::new((self.base_address + 0x000) as *const WriteOnly<...>) }
    }
    const fn tasks_hfclkstop(&self) -> ...
    ...
}
```

And a test named `ClockRegisters` that checks if any 2 fields (fields don't need to be ordered by offset in the macro input) are overlapping and fails with the following error message: `<prev_field> and <next_field> overlap in ClockRegisters`.

The main difference with what is currently done in Tock (and what this PR continues to do) is that accessing a register is done through a constant function call rather than a field access:
```rust
// Instead of
let regs = &*self.registers;
regs.intenset.write(...)
// It is
let regs = self.registers;
regs.intenset().write(...)
```

Also, initializing the registers is different and has a different type:
```rust
// Instead of
const CLOCK_BASE: StaticRef<ClockRegisters> =
    unsafe { StaticRef::new(0x40000000 as *const ClockRegisters) };
// It is
const CLOCK_BASE: ClockRegisters = ClockRegisters {
    base_address: 0x40000000,
};
```

Maybe this approach doesn't make sense, but it seems to me more adapted to Rust. The assembly should be exactly the same. So it's mostly a difference of APIs:
- As input, we provide for each register: name, type, offset. Without the need to mention and name padding (which is the only significant difference). Overlapping fields are still detected.
- As output, we get access to the register by calling a constant function instead of using a field.

Is there a reason I'm missing why structs are used in Tock to represent registers?

bradjc commented:
@ia0 I think the answer to your question from Tock's point of view is based on this generated piece of code:

```rust
const fn tasks_hfclkstart(&self) -> StaticRef<WriteOnly<u32, Control::Register>> {
    unsafe { StaticRef::new((self.base_address + 0x000) as *const WriteOnly<...>) }
}
```

Since a macro is creating that, chip drivers are using `unsafe` without the string "unsafe" appearing anywhere in the code. This type of hiding is pretty opposite of Tock's goals. With the struct approach we can be explicit that the unsafe operation is casting the MMIO pointer to a specific struct. Assuming that a developer has check that that address is correct, and the struct is correct, then that is memory safe. It also is grepable and easier to audit.

gendx commented:
@bradjc I briefly mentioned the reason in the issue https://github.com/tock/tock/issues/1400:

> Due to the difficulty of generating identifiers with macro_rules (https://stackoverflow.com/questions/27415011/can-a-rust-macro-create-new-identifiers), one still needs to spell out a name for each padding field, instead of using a generic name such as PADDING.

According to the StackOverflow thread, generating unique identifiers isn't an easy task with Rust macros, due to a lack of built-in support in the compiler. There seem to be a `mashup` crate that could allow that, but I'm not sure it's worth adding this extra dependency, nor how intuitive the syntax will end up.

If you thing it's worth it, the syntax could be `0x060 => @RESERVED(_reserved0)` instead of `0x060 => _reserved0`. I personally find that the extra verbosity is not worth it, but it wouldn't be hard to update the macros to support that.

gendx commented:
Ok, I'll figure out how to disable the check to allow `super::super::*` in this case.

ia0 commented:
@bradjc I see, it makes a lot of sense. Since the question is only about making the use of unsafe apparent (the macro definition has to be reviewed in any circumstances due to the [scope of unsafe][1]), it is easy to force the user to pay attention by forcing them to mention the `unsafe` keyword where something has to be checked.

First, the input of the macro must be checked (register names, offsets, and types):
```rust
define_registers! {
    unsafe struct ClockRegisters {
        tasks_hfclkstart: WriteOnly<u32, Control::Register> = 0x000,
        tasks_hfclkstop: WriteOnly<u32, Control::Register> = 0x004,
        tasks_lfclkstart: ReadWrite<u32, Control::Register> = 0x008,
```

Second, the base address has to be checked:
```rust
const CLOCK_BASE: ClockRegisters = ClockRegisters {
    base_address: unsafe { BaseAddress::new(0x40000000) },
};
```

Here is how the implementation looks like (omitting the glue):
```rust
// libraries/tock-register-interface/src/lib.rs
#[derive(Clone, Copy)]
pub struct BaseAddress {
    base_address: usize,
}
impl BaseAddress {
    pub const unsafe fn new(base_address: usize) -> BaseAddress {
        BaseAddress { base_address }
    }
    pub const fn offset(&self, offset: usize) -> usize {
        self.base_address + offset
    }
}

// libraries/tock-register-interface/src/macros.rs
#[macro_export]
macro_rules! define_registers {
    (unsafe struct $Name: ident { $($name: ident: $type: ty = $offset: expr),*$(,)? }) => {
        #[derive(Clone, Copy)]
        struct $Name { base_address: $crate::BaseAddress }
        impl $Name {
            $(const fn $name(&self) -> StaticRef<$type> {
                unsafe { StaticRef::new(self.base_address.offset($offset) as *const $type) }
            })*
        }
        // Omitting the test generation which is identical to before.
    };
}
```

[1]: https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html

rajivr commented:
Firstly this is really awesome! :-) 

I just tested this. I am running into two issues.

Following form compiles correctly.

```
register_structs!{
    UsbRegisters [
        (0x000 => usb_inten: ReadWrite<u32, USB_INTEN::Register>),
	...
	(0xXXX => @END),
    ]
}
```

However, when generating code from `svd2regs.py`, two things happen.

1. When comments are included (e.g., `/// USB Register Map`), it fails to compile. `svd2regs.py` generates comments by default.

```
register_structs!{
    /// USB Register Map
    UsbRegisters [
        (0x000 => usb_inten: ReadWrite<u32, USB_INTEN::Register>),
	...
	(0xXXX => @END),
    ]
}
````

2. It seems like `[` and `{` is being interchanged by `svd2regs.py`. Following form is generated that fails.

```
register_structs![
    /// USB Register Map
    UsbRegisters {
        (0x000 => usb_inten: ReadWrite<u32, USB_INTEN::Register>),
	...
	(0xXXX => @END),
    }
]
```

ia0 commented:
Another point against using structs is that it relies on `repr(C)`. This is probably not an issue in practice, but it's one more dependency that can go wrong. This could be fixed in the current PR by checking, in the generated test, that the field offsets in the output struct are actually equal to their offset in the macro input.

gendx commented:
@rajivr Thanks for the feedback! I updated the syntax to use curly braces everywhere in the macro, this should avoid confusion with square brackets. The macro should also accept comments before structs.

@ia0 That would be a great idea, but it seems that the `offset_of` macro in Rust's standard library is still under discussion: https://internals.rust-lang.org/t/pre-rfc-add-a-new-offset-of-macro-to-core-mem/9273
I'd wait for it to settle down before integrating it here.
In the meantime, there is a sizeof check for the whole struct, besides sizeof and alignof checks for individual fields, so if `repr(C)` somehow yields something unexpected it's likely to be caught by these tests.

If someone is more expert than me on Bash and can figure out a way to easily white-list `src/macro.rs` for wildcard imports **only** in `libraries/tock-register-interface/` that would be awesome!

gendx commented:
@bradjc I updated a sentence that I didn't find very clear in the readme.

ia0 commented:
@gendx Right, the existing checks make it already very unprobable that the offsets would be wrong. Adding such checks is unnecessary.

rajivr commented:
> @rajivr Thanks for the feedback! I updated the syntax to use curly braces everywhere in the macro, this should avoid confusion with square brackets. The macro should also accept comments before structs.

Thanks for the update. I am out all day today. I'll try to test it later this evening or tomorrow.

bradjc commented:
> @bradjc I updated a sentence that I didn't find very clear in the readme.

Is it not true that gaps become arrays? Does this make an array of `cr` that is 8 entries long?

```rust
register_structs! {
    Registers {
        (0x000 => cr: ReadWrite<u8, Control::Register>),
        (0x008 => s: ReadOnly<u8, Status::Register>),
    }
}
```

In either case the readme should specify what happens. I thought based on how reserved work that it would create an array.

ppannuto commented:
I don't think gaps can automatically become arrays, because the array would
need a name ("_reservered2" or whatever), and there isn't a way to create
names in macros right now

On Tue, Oct 22, 2019 at 9:54 AM Brad Campbell <notifications@github.com>
wrote:

> @bradjc <https://github.com/bradjc> I updated a sentence that I didn't
> find very clear in the readme.
>
> Is it not true that gaps become arrays? Does this make an array of cr
> that is 8 entries long?
>
> register_structs! {
>     Registers {
>         (0x000 => cr: ReadWrite<u8, Control::Register>),
>         (0x008 => s: ReadOnly<u8, Status::Register>),
>     }
> }
>
> In either case the readme should specify what happens. I thought based on
> how reserved work that it would create an array.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tock/tock/pull/1410?email_source=notifications&email_token=AACS3XT7E6CGZV6BH3O5NRDQP4V3NA5CNFSM4JBAJEHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEB6OHUI#issuecomment-545055697>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACS3XWBOGV3HPZVXZ77HHLQP4V3NANCNFSM4JBAJEHA>
> .
>

gendx commented:
No, your example would try to create:

```rust
#[repr(C)]
struct Registers {
    cr: ReadWrite<u8, Control::Register>,
    s: ReadOnly<u8, Status::Register>,
}
```

and then unit tests which would essentially do:

```rust
let mut sum = 0;
assert_eq(sum, 0);
sum += size_of::<u8>();
assert_eq(sum, 8);
sum += size_of::<u8>();
...
```

So the assert would break, but that would be detected by CI only after https://github.com/tock/tock/pull/1393 is merged.

It is however true that gaps becomes arrays of `u8` (ideally it should be something like `NoAccess<u8>` though), but gaps require an explicit identifier such as `_reserved0`.

rajivr commented:
I tested the latest patch. It look good! :+1:

bradjc commented:
So doing this:

https://github.com/tock/tock/blob/69c3a5d5583468a7775c5f193e7d05b7b1769755/chips/sam4l/src/acifc.rs#L84

would be:

```rust
register_structs! {
    Registers {
        (0x000 => [cr: ReadWrite<u8, Control::Register>; 8]),
        (0x008 => s: ReadOnly<u8, Status::Register>),
    }
}
```

gendx commented:
You can use any type as a "register", including an array type - in this case the type is `[ReadWrite<u32, WindowConfiguration::Register>; 4]`.

So here you'd need something like:

```rust
register_structs! {
AcifcRegisters {
    (0x000 => ctrl: ReadWrite<u32, Control::Register>),
    ...
    (0x038 => _reserved2),
    (0x080 => confw: [ReadWrite<u32, WindowConfiguration::Register>; 4]),
    (0x090 => _reserved3),
    (0x0D0 => conf: [ReadWrite<u32, ACConfiguration::Register>; 8]),
    (0xXXX => @END),
}
}
```

It's actually similar to the regular syntax for structs, apart from the explicit addresses, and the implicit type for gaps.

I can add this case to the examples if it makes it clearer.

bradjc commented on tools/run_cargo_fmt.sh:
```diff
@@ -64,7 +64,7 @@ done
 for f in $(find . | grep Cargo.toml); do
 	pushd $(dirname $f) > /dev/null
-	if $(git grep -q 'use .*\*;'); then
+	if $(git grep -q 'use .*\*;' -- ':!src/macros.rs'); then
```
Why does macros.rs need to be excluded?

I prefer to trust in rustfmt, but there are ways to be very specific about what fustfmt should ignore.

gendx replied:
The confusion part here is that this `run_rust_fmt.sh` script also handles looking for wildcard import patterns (the part that I changed here), which has nothing to do with formatting. I'd suggest to split the script in 2 files to make this clearer.

gendx replied:
Moved that part to a standalone script.

bradjc commented:
Ok gotcha, and thanks for adding that to the readme.

One last question: so only entries with a single name (`(0x090 => _reserved3)`) automatically become arrays, or only things that start with `_reserved`? Put another way, if someone specifies the space as `(0x090 => padding)` will that work?

gendx commented:
The macro doesn't actually check the identifier, you can use whatever you want, including "padding". I would say that using `_reservedN` would be more of a style convention within Tock.

Note that since such fields are never used (or at least they shouldn't!), not pre-pending them with an underscore will lead the compiler to complain about dead code. And you cannot reuse twice the same identifier in a struct, so if you have multiple gaps, you'd need to name them something like `padding0`, `padding1`, etc. The simplest is probably to use `_reservedN` (which is also output by `svd2regs.py`).

bradjc commented:
But it is based on just having an identifier? I'm still trying to understand why in one case an array is automatically created and in my earlier case an array is not automatically created and instead a gap is introduced.

As for the naming, I just want to make sure I understand so I can help users if they have questions in the future.

gendx commented:
The syntax is defined as follows in the macro:

- `($offset_start:expr => $field:ident: $ty:ty)` yields `$field: $ty` in the struct.
- `($offset_start:expr => $padding:ident)` yields `$padding: [u8; $offset_end - $offset_start]` in the struct (where `$offset_end:expr` is obtained from the next field).

So it is based on having a type (separated by a colon) or not. In all cases an identifier is required, as there's no easy way to generate one in a Rust macro (as discussed in the issue).

If you find that confusing, I could make the syntax more explicit, e.g. `0x009 => @PADDING(_reserved0)`.

bradjc commented:
Ok I think I understand now. Please double check my updates to the readme, but I think they are consistent with the code and should give people enough detail to effectively use the register_structs macro.

andre-richter commented:
That sounds incredibly useful. Thanks!

One thing I noticed is that the README is kind of a roller-coaster where you first think that the no.1 selling point of this addition, naemly automatic gap calculation, isn't there. Readers will first stumble upon this sentence:

```
Any gap when defining the registers will **not** be preserved when the register struct
is created by the macro. The macro also generates a unit test which will fail if a gap
is introduced.
```

I read that and was like, bummer, I still have to manually fill. Only the next few sentences explain it, but still its not 100% clear.

I would put it something like:
"Gaps when defining the registers must be explicitly annotated (by convention using `_reservedN`), the macro will then automatically take care of calculating the gap size and inserting a respective filler struct.

gendx commented:
@andre-richter I think the original formulation in the documentation had a more direct explanation. Thanks for your proposal, this sounds clear!

gendx commented:
I added an example of unintentional gap. I hope that this is clear enough (and that the discussion on https://github.com/tock/tock/pull/1393 will settle down).

alevy commented:
So just waiting on @bradjc to approve I think

ia0 commented:
FYI: I created #1439 to continue discussion on avoiding encoding the register name to offset mapping in the field offsets of a `repr(C)` struct. This should not block merging this PR which is already improving the situation.

alevy commented:
bors r+

gendx commented:
Looks like this is still not merged by bors.

alevy commented:
bors r+
